src/JustSaying/Messaging/Channels/Multiplexer/RoundRobinQueueMultiplexer.cs

```diff
@@ -0,0 +1,179 @@
+        }
+
+        public void ReadFrom(ChannelReader<IQueueMessageContext> reader)
+        {
+            if (reader == null) throw new ArgumentNullException(nameof(reader));
+
+            _readersLock.Wait(_stoppingToken);
+
+            try
+            {
+                _readers.Add(reader);
+            }
+            finally
+            {
+                _readersLock.Release();
+            }
+
+            async Task OnReaderCompletion()
+            {
+                await reader.Completion.ConfigureAwait(false);
+                RemoveReader(reader);
+            }
+
+            _ = OnReaderCompletion();
+        }
+
+        private void RemoveReader(ChannelReader<IQueueMessageContext> reader)
+        {
+            _logger.LogInformation("Received notification to remove reader from multiplexer inputs");
+
+            _readersLock.Wait(_stoppingToken);
+
+            try
+            {
+                _readers.Remove(reader);
+            }
+            finally
+            {
+                _readersLock.Release();
+            }
+        }
+
+        public Task Run(CancellationToken stoppingToken)
+        {
+
+            if (!_started)
+            {
+                lock (_startLock)
+                {
+                    if (!_started)
+                    {
+                        _stoppingToken = stoppingToken;
+                        _completion = RunImpl();
+                        _started = true;
+                    }
+                }
+            }
+
+            return _completion;
+        }
+
+        private async Task RunImpl()
+        {
+            await Task.Yield();
+
+            _logger.LogInformation(
+                "Starting up channel multiplexer with a queue capacity of {Capacity}",
+                _channelCapacity);
+
+            ChannelWriter<IQueueMessageContext> writer = _targetChannel.Writer;
+            while (true)
+            {
+                await _readersLock.WaitAsync(_stoppingToken).ConfigureAwait(false);
+
+                _stoppingToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    if (_readers.Count < 1)
+                    {
+                        _logger.LogInformation("All writers have completed, terminating multiplexer");
+                        writer.Complete();
+                        break;
+                    }
+
+                    bool any = false;
+                    foreach (ChannelReader<IQueueMessageContext> reader in _readers)
+                    {
+                        if (reader.TryRead(out IQueueMessageContext message))
+                        {
+                            any = true;
+                            await writer.WriteAsync(message, _stoppingToken).ConfigureAwait(false);
+                        }
+                    }
+
+                    if (!any)
+                    {
+                        await Task.Delay(5).ConfigureAwait(false);
```

Comment:
This doesn't seem ideal, we'd be spamming the theadpool and adding timer queue pressure unnecessarily when we should be idle.
Perhaps something along the lines of:
```
await Task.WhenAny(_readers.Select(r => r.WaitToReadAsync())).ConfigureAwait(false);
```
Not sure if you'd want to cancel the others when one returns.

Reply:
Yeah this is a good point. If we didn't cancel then I'm not sure if that would cause any issues later on. In my linqpad experiments the call to `WaitToReadAsync` doesn't block any following `TryRead`'s or anything so I think it should be ok. 

The only other thing is that we need to async/await in the Select as WhenAny has no overload for ValueTask - not sure there's any way around this?

Reply:
Ah, great point. We could use `AsTask()`, they will almost always be async waits because we've `TryRead` returned false on all readers by that point.

Reply:
Yeah that's a great point - this would ease pressure on the thread pool quite a bit I'd think. Will make it so!

Reply:
Perhaps a minor drawback of this strategy is that it would no longer be possible to dynamically add/remove  readers to the multiplexer. It's not a use case we currently support via the API, so it's not a big deal right now. Each time we go around the loop, we're using the current list of readers so it allows them to be added/removed on the fly.

Reply:
This is no longer an issue as this multiplexer has been replaced by a new MergingMultiplexer